nautobot/extras/migrations/0057_collect_roles_from_related_apps_roles.py

```diff
@@ -61,22 +61,46 @@ def bulk_create_roles(apps, roles_to_create, content_types):
     """Bulk create role and set its contenttypes"""
     Role = apps.get_model("extras", "Role")
     ContentType = apps.get_model("contenttypes", "ContentType")
+    ObjectChange = apps.get_model("extras", "ObjectChange")
+
+    old_role_ct = None
+    new_role_ct = None
+
+    if hasattr(roles_to_create, "model"):
+        old_role_ct = ContentType.objects.get_for_model(roles_to_create.model)
+        # No sense fetching this if we don't have an old role content type
+        new_role_ct = ContentType.objects.get_for_model(Role)
 
     # Exclude roles whose names are already exists from bulk create.
     existing_roles = Role.objects.values_list("name", flat=True)
     roles_instances = [
-        Role(
-            name=data.name,
-            slug=data.slug,
-            description=data.description,
-            color=getattr(data, "color", color_map["default"]),
-            weight=getattr(data, "weight", None),
+        (
+            data,
+            Role(
+                name=data.name,
+                slug=data.slug,
+                description=data.description,
+                color=getattr(data, "color", color_map["default"]),
+                weight=getattr(data, "weight", None),
+            ),
         )
         for data in roles_to_create
         if data.name not in existing_roles
     ]
     if roles_instances:
-        Role.objects.bulk_create(roles_instances, batch_size=1000)
+        for old_role, new_role in roles_instances:
+            # TODO: We need to handle new role collisions
+            new_role.save()
+            if old_role_ct:
+                # Move over existing object change records to the new role we created
+                ObjectChange.objects.filter(changed_object_type=old_role_ct, changed_object_id=old_role.pk).update(
+                    changed_object_type=new_role_ct, changed_object_id=new_role.pk
+                )
+
+    # This is for all of the change records for roles which no longer exist
+    if old_role_ct:
+        # TODO: Should we null out the changed_object_id?
```

Comment:
I don't think we do for object deletes in general?

Reply:
We don't which is why I didn't do it here and a UUID is a UUID and even if it were in the same ContentType the likelihood you'll re-use a UUID is staggeringly small.